### ISSUE TITLE
ERR V3Sub: EnsurePolling unknown device error

### ISSUE THREAD

**Issue by mike-pisman** [matrix-org/sliding-sync#364]
Hi, 

When trying to login with element-x client, I get stuck on "Setting up your account" screen.

I also get a SQL error in the sliding-sync container:

```
ERR V3Sub: EnsurePolling unknown device error="sql: no rows in result set" device_id=******* user_id=@myusername:matrix.mydomain.com
```

<details>
  <summary>Full log</summary>

```
Sync v3 [0.99.11] (a86e01a)
Debug=false LogLevel= MaxConns=0
2023/11/04 22:39:11 OK   20230728114555_device_data_drop_id.sql (16.13ms)
2023/11/04 22:39:11 OK   20230802121023_device_data_jsonb.go (39.45ms)
2023/11/04 22:39:11 OK   20230814183302_cbor_device_data.go (19.01ms)
2023/11/04 22:39:11 OK   20230822180807_bogus_snapshot_cleanup.go (23.41ms)
2023/11/04 22:39:11 OK   20230913120537_events_missing_previous.sql (14.72ms)
2023/11/04 22:39:11 goose: successfully migrated database to version: 20230913120537
22:39:11 INF creating handler
22:39:11 INF retrieved global snapshot from database
22:39:11 INF listening on 0.0.0.0:8008
22:39:11 INF StartV2Pollers num_devices=0 num_fail_decrypt=0
22:39:11 INF StartV2Pollers finished
22:39:23 INF Received connection from unknown access token, querying with homeserver
22:39:23 INF Received connection from unknown access token, querying with homeserver
22:39:23 INF EnsurePolling: new request device_id=*** user_id=@myusername:matrix.mydomain.com
22:39:23 ERR V3Sub: EnsurePolling unknown device error="sql: no rows in result set" device_id=*** user_id=@myusername:matrix.mydomain.com
22:39:23 INF EnsurePolling: preprocessing done device_id=*** user_id=@myusername:matrix.mydomain.com
```

</details>

I use:
Postgresql: 16.0
Dendrite: 13.3

I have tried to create a new Postgresql database for Sliding-sync, but the issue persists. Could this be caused by my matrix server/database?

I have done a bit of digging around:

The error comes from this code block:
```go
accessToken, since, err := h.v2Store.TokensTable.GetTokenAndSince(p.UserID, p.DeviceID, p.AccessTokenHash)
if err != nil {
    log.Err(err).Msg("V3Sub: EnsurePolling unknown device")
    sentry.CaptureException(err)
    return
}
```

And the `GetTokenAndSince` function definition:
```go
func (t *TokensTable) GetTokenAndSince(userID, deviceID, tokenHash string) (accessToken, since string, err error) {
	var encToken, gotUserID, gotDeviceID string
	query := `SELECT token_encrypted, since, user_id, device_id
	FROM syncv3_sync2_tokens JOIN syncv3_sync2_devices USING (user_id, device_id)
	WHERE token_hash = $1;`
	err = t.db.QueryRow(query, tokenHash).Scan(&encToken, &since, &gotUserID, &gotDeviceID)
	if err != nil {
		return
	}
	if gotUserID != userID || gotDeviceID != deviceID {
		err = fmt.Errorf(
			"token (hash %s) found with user+device mismatch: got (%s, %s), expected (%s, %s)",
			tokenHash, gotUserID, gotDeviceID, userID, deviceID,
		)
		return
	}
	accessToken, err = t.decrypt(encToken)
	return
}
```

So I tried to run the query myself
```sql
SELECT token_encrypted, since, user_id, device_id FROM syncv3_sync2_tokens JOIN syncv3_sync2_devices USING (user_id, device_id);
```
And I get the row with my `token_encrypted`, `user_id`, `device_id`. The `since` column is empty, though I'm not sure if it matters. 

So I assume the issue lies in the `access_token`, as the log states
```
Received connection from unknown access token
```

How can I troubleshoot the problem further, thanks.

## Update

After restarting the server multiple times, something happened and I managed to sign in on element-x-ios client. However, I am not able to verify the session.

```
ERR failed to upsert device data error="pq: invalid byte sequence for encoding \"UTF8\": 0xa6" user=@***:matrix.mydomain.com 
ERR Poller: parseE2EEData returned an error error="pq: invalid byte sequence for encoding \"UTF8\":
```

## Update 2

I have deleted and recreated database with UTF8 format, and managed to connect element-x-ios client. Some message history is hidden, but it seems to work. Still no idea what has changed, that I managed to connect.

I am still not able to connect the element-x-android client. Here is the error:
```
 TRC liveUpdate: no response data yet; blocking device
```

**Comment by kegsay** [comment]
It sounds like the user IDs you're using are not valid UTF8. This will break in weird and wonderful ways. Even if you can get the proxy to work with them, you'll face an uphill struggle with rust SDK as there is validation done there as well.

My guess is that your manual query works because you aren't parsing the user/device as UTF8 (it's just a JOIN). This is subtly different to what the proxy does, which will be to do a literal query with the string e.g `WHERE user_id=$1 AND device_id=$2`.

**Comment by mike-pisman** [comment]
@kegsay, thanks for the reply.

I just checked again and both element-x clients(ios and android) just magically worked. I managed to sign in and verify the devices. 
I have not changed anything, and not sure what caused or resolved the issues... 
The only problem now is some rooms created with sliding-sync show up as Empty rooms and the messages are hidden, perhaps this is because the other users are not verified properly.
